components/patient-data/rest/src/test/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImplTest.java

```diff
@@ -0,0 +1,355 @@
+        assertEquals("[]", response.getEntity());
+    }
+
+    @Test
+    public void getPatientsPerformsCorrectlyOnePatientRecordByEidNoneById()
+        throws ComponentLookupException, QueryException, IOException
+    {
+        final List<Object> eidList = ImmutableList.<Object>of(EID1);
+
+        final Request request = mock(Request.class);
+        doReturn(request).when(this.container).getRequest();
+        doReturn(eidList).when(request).getProperties("eid");
+        doReturn(Collections.emptyList()).when(request).getProperties("id");
+
+        final Query query = mock(DefaultQuery.class);
+        doReturn(this.patient1JSON).when(this.patient1).toJSON();
+        doReturn(query).when(this.qm).createQuery(Matchers.anyString(), Matchers.anyString());
+        doReturn(ImmutableList.<Object>of(ID_1)).when(query).execute();
+
+        doReturn(this.patient1).when(this.repository).get(ID_1);
+        doReturn(ID_1).when(this.patient1).getId();
+
+        final Response response = this.component.fetchPatients();
+
+        assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
+
+        final String expectedStr = new JSONArray().put(new JSONObject().put("id", ID_1).put("links", this.uriList))
+            .toString();
+        @SuppressWarnings("unchecked")
+        final List<String> expected = (List<String>) (OBJECT_MAPPER.readValue(expectedStr, List.class));
+        @SuppressWarnings("unchecked")
+        final List<String> actual = (List<String>) (OBJECT_MAPPER.readValue((String) response.getEntity(), List.class));
+        assertTrue(expected.containsAll(actual) && actual.containsAll(expected));
+        assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
+    }
+
+    @Test
+    public void getPatientsPerformsCorrectlySeveralPatientRecordsAccessToAll() throws ComponentLookupException,
+        QueryException, IOException
+    {
+        final List<Object> eidList = new ArrayList<>();
+        eidList.add(EID1);
+        eidList.add(EID2);
+
+        final List<String> idList = new ArrayList<>();
+        idList.add(ID_3);
+
+        final Request request = mock(Request.class);
+        doReturn(request).when(this.container).getRequest();
+        doReturn(eidList).when(request).getProperties("eid");
+        doReturn(idList).when(request).getProperties("id");
+
+        final Query query = mock(DefaultQuery.class);
+        when(this.patient1.toJSON()).thenReturn(this.patient1JSON);
+        when(this.patient2.toJSON()).thenReturn(this.patient2JSON);
+        when(this.patient3.toJSON()).thenReturn(this.patient3JSON);
+
+        when(this.qm.createQuery(Matchers.anyString(), Matchers.anyString())).thenReturn(query);
+        when(query.execute()).thenReturn(ImmutableList.<Object>of(ID_1, ID_2));
+        doReturn(this.patient1).when(this.repository).get(ID_1);
+        doReturn(this.patient2).when(this.repository).get(ID_2);
+        doReturn(this.patient3).when(this.repository).get(ID_3);
```

Comment:
You can override the behavior when needed, the last mocking call is the one that will be taken into account.